python/lsst/meas/algorithms/stamps.py

```diff
@@ -0,0 +1,351 @@
+    fitsPrimary.createEmpty()
+    fitsPrimary.writeMetadata(metadata)
+    fitsPrimary.closeFile()
+
+    # add all pixel data optionally writing mask and variance information
+    for i, stamp in enumerate(stamp_ims):
+        metadata = PropertySet()
+        metadata.update({'EXTVER': i, 'EXTNAME': 'IMAGE'})
+        stamp.getImage().writeFits(filename, metadata=metadata, mode='a')
+        if write_mask:
+            metadata = PropertySet()
+            metadata.update({'EXTVER': i, 'EXTNAME': 'MASK'})
+            stamp.getMask().writeFits(filename, metadata=metadata, mode='a')
+        if write_variance:
+            metadata = PropertySet()
+            metadata.update({'EXTVER': i, 'EXTNAME': 'VARIANCE'})
+            stamp.getVariance().writeFits(filename, metadata=metadata, mode='a')
+    return None
+
+
+def readFitsWithOptions(filename, stamp_factory, options):
+    """Read stamps from FITS file, allowing for only a
+    subregion of the stamps to be read.
+
+    Parameters
+    ----------
+    filename : `str`
+        A string indicating the file to read
+    stamp_factory : classmethod
+        A factory function defined on a dataclass for constructing
+        stamp objects a la `lsst.meas.alrogithm.Stamp`
+    options : `PropertySet`
+        A collection of parameters.  If certain keys are available
+        (``llcX``, ``llcY``, ``width``, ``height``), a bounding box
+        is constructed and passed to the ``FitsReader`` in order
+        to return a sub-image.
+
+    Returns
+    -------
+    stamps, metadata : `list` of dataclass objects like `Stamp`, PropertySet
+        A tuple of a list of `Stamp`-like objects and a collection of metadata.
+    """
+    # extract necessary info from metadata
+    metadata = afwFits.readMetadata(filename, hdu=0)
+    nStamps = metadata["N_STAMPS"]
+    # check if a bbox was provided
+    kwargs = {}
+    if options and options.exists("llcX"):
+        llcX = options["llcX"]
+        llcY = options["llcY"]
+        width = options["width"]
+        height = options["height"]
+        bbox = Box2I(Point2I(llcX, llcY), Extent2I(width, height))
+        kwargs["bbox"] = bbox
+    stamp_parts = {}
+    idx = 1
+    while len(stamp_parts) < nStamps:
+        md = afwFits.readMetadata(filename, hdu=idx)
+        if md['EXTNAME'] in ('IMAGE', 'VARIANCE'):
+            reader = afwImage.ImageFitsReader(filename, hdu=idx)
+        elif md['EXTNAME'] == 'MASK':
+            reader = afwImage.MaskFitsReader(filename, hdu=idx)
+        else:
+            raise ValueError(f"Unknown extension type: {md['EXTNAME']}")
+        stamp_parts.setdefault(md['EXTVER'], {})[md['EXTNAME'].lower()] = reader.read(**kwargs)
+        idx += 1
+    # construct stamps themselves
+    stamps = []
+    meta_dict = metadata.toDict()
+    for k in stamp_parts:
+        maskedImage = afwImage.MaskedImageF(**stamp_parts[k])
+        # Assume the value of EXTVER is the index into the metadata
```

Comment:
EXTVER starts at 1 and not 0. Does that matter?

Reply:
Oh, good catch.  I think it may.  Let me look.

Reply:
I just made EXTVER start at one so it would line up with the extension number.  I think that's a bad idea since it will also be used as an index into an array of values for building the stamps.  I'm changing to make it zero based

Reply:
I have no idea if 0 is allowed for EXTVER. You may have to read the standard a bit harder to see if 0 is allowed but one option is to create the file starting at 0 and see if fitsverify is happy with it.

Reply:
Will do

Reply:
OK.  The standard specifically says it must be > 0.